rs/registry/canister/canister/canister.rs

```diff
@@ -185,60 +185,67 @@ ic_nervous_system_common_build_metadata::define_get_build_metadata_candid_method
 
 #[export_name = "canister_query get_changes_since"]
 fn get_changes_since() {
-    // Only self-authenticating and anonymous principals are allowed to call us.
-    const ALLOWED_CALLER_CLASSES: [Result<PrincipalIdClass, String>; 2] = [
-        Ok(PrincipalIdClass::SelfAuthenticating),
-        Ok(PrincipalIdClass::Anonymous),
-    ];
-    let class = caller().class();
-    if !ALLOWED_CALLER_CLASSES.contains(&class) {
-        let response = RegistryGetChangesSinceResponse {
-            error: Some(RegistryError {
-                code: Code::Authorization as i32,
-                reason: format!(
+    fn main() -> Result<RegistryGetChangesSinceResponse, (Code, String)> {
+        // Authorize: Only self-authenticating and anonymous principals are allowed to call us.
+        const ALLOWED_CALLER_CLASSES: [Result<PrincipalIdClass, String>; 2] = [
+            Ok(PrincipalIdClass::SelfAuthenticating),
+            Ok(PrincipalIdClass::Anonymous),
+        ];
+        let class = caller().class();
+        if !ALLOWED_CALLER_CLASSES.contains(&class) {
+            return Err((
+                Code::Authorization,
+                format!(
                     "Caller must be self-authenticating, or anonymous (but was {:?}).",
                     class,
                 ),
-                key: vec![],
-            }),
-            ..Default::default()
+            ));
+        }
+
+        // Parse request.
+        let request = match deserialize_get_changes_since_request(arg_data()) {
+            Ok(ok) => ok,
+            Err(error) => {
+                return Err((Code::MalformedMessage, error.to_string()));
+            }
         };
+        let version = request;
 
-        let response: Vec<u8> =
-            serialize_get_changes_since_response(response).expect("Error serializing response");
+        // All requirements met. Proceed with "real work".
 
-        reply(&response);
-        return;
```

Comment:
This is duplicated later. Getting rid of this duplication is one of the goals of this PR.